services/typescript-service.ts

```diff
@@ -0,0 +1,316 @@
+	}
+
+	private getTypeScriptCompiler(): IFuture<ITypeScriptCompilerSettings> {
+		return ((): ITypeScriptCompilerSettings => {
+			let tempTscDirectory: string;
+			if (!this.hasInstalledTsc) {
+				tempTscDirectory = this.createTempDirectoryForTsc().wait();
+				let pluginToInstall: INpmDependency = {
+					name: TypeScriptService.TYPESCRIPT_MODULE_NAME,
+					version: TypeScriptService.DEFAULT_TSC_VERSION,
+					installTypes: false
+				};
+				this.$npmService.install(tempTscDirectory, pluginToInstall).wait();
+				this.hasInstalledTsc = true;
+			}
+
+			// Get the path to tsc
+			let typeScriptModuleFilePath = path.join(tempTscDirectory, "node_modules", TypeScriptService.TYPESCRIPT_MODULE_NAME);
+			let typeScriptCompilerPath = path.join(typeScriptModuleFilePath, "lib", "tsc");
+			let typeScriptCompilerVersion = this.$fs.readJson(path.join(typeScriptModuleFilePath, "package.json")).wait().version;
+
+			return { pathToCompiler: typeScriptCompilerPath, version: typeScriptCompilerVersion };
+		}).future<ITypeScriptCompilerSettings>()();
+	}
+
+	private runTranspilation(projectDir: string, typeScriptCompilerPath: string, options?: { typeScriptCommandsFilePath?: string, compilerOptions?: ITypeScriptCompilerOptions }): IFuture<string> {
+		return ((): string => {
+			options = options || {};
+			let startTime = new Date().getTime();
+			let params = [typeScriptCompilerPath];
+			if (options.typeScriptCommandsFilePath) {
+				params.push("@" + options.typeScriptCommandsFilePath);
+			} else if (options.compilerOptions) {
+				params = _.concat(params, this.getTypeScriptCompilerOptionsAsArguments(options.compilerOptions));
+			}
+
+			let output = this.$childProcess.spawnFromEvent(process.argv[0], params, "close", { cwd: projectDir }, { throwError: false }).wait();
+			let exitcode = output.exitCode;
+			let compilerOutput = output.stderr || output.stdout;
+
+			// EmitReturnStatus enum in https://github.com/Microsoft/TypeScript/blob/8947757d096338532f1844d55788df87fb5a39ed/src/compiler/types.ts#L605
+			if (exitcode === 0 || exitcode === 2 || exitcode === 3) {
+				let endTime = new Date().getTime();
+				let time = (endTime - startTime) / 1000;
+
+				this.$logger.out(`${os.EOL}Success: ${time.toFixed(2)}s${os.EOL}Done without errors.`.green);
+			} else {
+				let compilerMessages = this.getCompilerMessages(compilerOutput);
+				this.logCompilerMessages(compilerMessages, compilerOutput);
+			}
+
+			return compilerOutput;
+		}).future<string>()();
+	}
+
+	private getCompilerMessages(compilerOutput: string): ITypeScriptCompilerMessages {
+		// Assumptions:
+		//   Level 1 errors = syntax errors - prevent JS emit.
+		//   Level 2 errors = semantic errors - *not* prevents JS emit.
+		//   Level 5 errors = compiler flag misuse - prevents JS emit.
+
+		let level1ErrorCount = 0,
+			level5ErrorCount = 0,
+			nonEmitPreventingWarningCount = 0;
+
+		let hasPreventEmitErrors = _.reduce(compilerOutput.split("\n"), (memo: any, errorMsg: string) => {
```

Comment:
you can narrow down the type of the argument `memo` in the lamda function as it is actually `boolean` instead of `any`